pkg/controller/configmap/configmap_controller.go

```diff
@@ -0,0 +1,129 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configmap
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"os"
+
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName   = "configmap"
+	configMapName    = "cco-trusted-ca"
+	configMapKeyName = "ca-bundle.crt"
+	ccoNamespace     = "openshift-cloud-credential-operator"
+)
+
+// Add creates a new ConfigMap controller and adds it to the manager.
+func Add(mgr manager.Manager) error {
+	return AddToManager(mgr, NewReconciler(mgr))
+}
+
+// NewReconciler returns a new reconcile.Reconciler
+func NewReconciler(mgr manager.Manager) *ReconcileConfigMap {
+	return &ReconcileConfigMap{
+		Client: mgr.GetClient(),
+		logger: log.WithField("controller", controllerName),
+		exit:   exitFunc,
+	}
+}
+
+// AddToManager adds a new Controller to mgr with r as the reconcile.Reconciler
+//func AddToManager(mgr manager.Manager, r *ReconcileConfigMap) error {
+func AddToManager(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		log.WithField("controller", controllerName).WithError(err).Error("Error creating controller")
+		return err
+	}
+
+	// Watch for changes to the configMap containing the trusted CAs
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return configMapCheck(e.MetaNew)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return configMapCheck(e.Meta)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return configMapCheck(e.Meta)
+		},
+	}
+	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func configMapCheck(cm metav1.Object) bool {
+	return cm.GetName() == configMapName && cm.GetNamespace() == ccoNamespace
+}
+
+var _ reconcile.Reconciler = &ReconcileConfigMap{}
+
+// ReconcileConfigMap reconciles a ConfigMap against the on-disk contents.
+type ReconcileConfigMap struct {
+	client.Client
+	logger            log.FieldLogger
+	configMapDataHash string
```

Comment:
This is persistent state across reconcile loops isn't it. That is interesting, we don't have a lot of examples of that. It looks risky for concurrency, but then we're only reconciling on one single config map. Looks like it should be ok to me.

Reply:
Yes, persistent across reconcile calls. And as you say, since we're really only watching one single configmap (openshift-cloud-credential-operator/cco-trusted-ca), this should be safe.